### ISSUE TITLE
QuantRepNCSPELAN4Chunk and remove_redundant_qdq_model

### ISSUE THREAD

**Issue by marcoslucianops** [levipereira/yolov9-qat#19]
Hi, thanks for your awesome work!

Is there any reason why `QuantRepNCSPELAN4Chunk` is disabled in the `models/quantize.py`?
Also, I saw there are removed parts from the `remove_redundant_qdq_model`. What's the reason?

**Comment by marcoslucianops** [comment]
I have a different implementation with modelopt.

EDIT: The results below is in a custom code for YOLO and YOLOv9-s

Enabling the `QuantRepNCSPELAN4Chunk` (with few changes to act as module), I get
```
[03/25/2025-10:29:56] [I] Throughput: 792.487 qps
[03/25/2025-10:29:56] [I] Latency: min = 1.24707 ms, max = 1.77734 ms, mean = 1.26009 ms, median = 1.25781 ms, percentile(90%) = 1.25977 ms, percentile(95%) = 1.26172 ms, percentile(99%) = 1.29688 ms
```
![Image](https://github.com/user-attachments/assets/3d936602-9cc5-444d-a35e-3a38e6b39e60)
![Image](https://github.com/user-attachments/assets/a3cef61e-d85d-4363-adb0-3ea300dea106)


Disabling the `QuantRepNCSPELAN4Chunk`, I get
```
[03/25/2025-10:28:16] [I] Throughput: 749.529 qps
[03/25/2025-10:28:16] [I] Latency: min = 1.31885 ms, max = 1.79688 ms, mean = 1.33277 ms, median = 1.33203 ms, percentile(90%) = 1.33203 ms, percentile(95%) = 1.33325 ms, percentile(99%) = 1.33594 ms
```
![Image](https://github.com/user-attachments/assets/da48a051-cb34-4f1c-add8-148322106640)
![Image](https://github.com/user-attachments/assets/a38fa087-f3f0-4dad-9d40-b809c5ffc889)

**Comment by levipereira** [comment]
Hi,
Regarding QuantRepNCSPELAN4Chunk : When implementing TensorRT(10.0), I noticed it was generating reformat operations due to unusual conversions between fp16/int8, which significantly increased processing time. This is why I disabled it.
I would recommend visualizing both the ONNX graph and the engine graph to compare them directly. 
I encountered similar challenges when quantizing YOLOv9-s, where numerous reformat operations actually degraded performance. The solution requires establishing proper rules and removing duplicated qat nodes to minimize these reformat operations.
These reformat operations are particularly problematic because TensorRT fails to fuse nodes (due to precision incompatibility or quantization duplication), resulting in additional reformat operations that hurt performance.
I began migrating to the ModelOpt approach, but the transition from pytorch-quantization to ModelOpt involves significant changes that require substantial implementation work.
May I ask how you're approaching model quantization with ModelOpt? Are you using ONNX Quantization or PyTorch Quantization? In my experience, ONNX Quantization yielded both lower performance and reduced accuracy.
I intend to resume this project soon, with a focus on GeneralistYOLO and YOLOe, which I believe represent the next evolutionary step for YOLO architectures

What GPU and Batch Size you are using? This is yolov9-s?
```
[03/25/2025-10:28:16] [I] Throughput: 749.529 qps
[03/25/2025-10:28:16] [I] Latency: min = 1.31885 ms, max = 1.79688 ms, mean = 1.33277 ms, median = 1.33203 ms, percentile(90%) = 1.33203 ms, percentile(95%) = 1.33325 ms, percentile(99%) = 1.33594 ms
```


Check this :
https://github.com/WongKinYiu/yolov9/issues/327#issuecomment-2155760078

**Comment by marcoslucianops** [comment]
I'm using torch quantization. Basically, I rewrote few things in the quantize and calibrate functions to use `max` with `mse` (from the modelopt functions), used `Quant` modules in `mtq.register`, and used some of your functions (with some rewrites).

My board is RTX 3090 TI and I'm using YOLOv9-s SiLU (from my custom training code), TensorRT 10.3.0 and CUDA 12.6.

I did few changes in the code and now I got (still trying to get it better):

- Enabling the QuantRepNCSPELAN4Chunk
  ```
  [03/26/2025-00:54:21] [I] Throughput: 811.43 qps
  [03/26/2025-00:54:21] [I] Latency: min = 1.21954 ms, max = 1.69531 ms, mean = 1.23088 ms, median = 1.22656 ms, percentile(90%) = 1.23438 ms, percentile(95%) = 1.23438 ms, percentile(99%) = 1.23828 ms
  ```
  ![Image](https://github.com/user-attachments/assets/aac1fa05-a2a6-4fd8-b232-2a9c34906b39)
  ![Image](https://github.com/user-attachments/assets/4eaed841-afaa-40a6-b56b-e4c127e1b575)

- Disabling the QuantRepNCSPELAN4Chunk
  ```
  [03/26/2025-00:56:11] [I] Throughput: 812.518 qps
  [03/26/2025-00:56:11] [I] Latency: min = 1.21753 ms, max = 2.42969 ms, mean = 1.22916 ms, median = 1.22485 ms, percentile(90%) = 1.23438 ms, percentile(95%) = 1.23438 ms, percentile(99%) = 1.23438 ms
  ```
  ![Image](https://github.com/user-attachments/assets/3a21c6d5-a0dc-4bcd-912a-1ae92f8951a8)
  ![Image](https://github.com/user-attachments/assets/70521eb4-c35a-495e-8632-94dfa01379f1)

- FP16 (for reference)
  ```
  [03/26/2025-00:58:06] [I] Throughput: 779.713 qps
  [03/26/2025-00:58:06] [I] Latency: min = 1.26465 ms, max = 1.75391 ms, mean = 1.2811 ms, median = 1.2793 ms, percentile(90%) = 1.28906 ms, percentile(95%) = 1.28906 ms, percentile(99%) = 1.29297 ms
  ```

**Comment by levipereira** [comment]
Great that you achieved this performance. I will test it later on TensorRT 10.3 to see if the reformat issue has been resolved.

Is your implementation open source, and if so, are you going to make it public?
 
I recommend testing with different batch sizes as that's where it really makes all the difference. 
This link shows a huge difference in how things work on the GPU in relation to batch size:
https://github.com/WongKinYiu/yolov9/issues/327#issuecomment-2155765527

In YOLOv9-C RELU performance gain I obtained was 30% with batch size 1 and almost 90% with batch size 12.
See this post:
https://github.com/WongKinYiu/yolov9/issues/327#issuecomment-2060285164"

**Comment by marcoslucianops** [comment]
I won't set it public for now. I'm using in some projects I have.

With the final changes, and removing the QuantRepNCSPELAN4 (same of your tests), it's a little fast

YOLOv9-c with batch-size 1

SiLU FP16
```
[03/27/2025-09:25:00] [I] Throughput: 431.76 qps
[03/27/2025-09:25:00] [I] Latency: min = 2.10425 ms, max = 3.25391 ms, mean = 2.31465 ms, median = 2.24219 ms, percentile(90%) = 2.57031 ms, percentile(95%) = 2.60547 ms, percentile(99%) = 2.79297 ms
```

ReLU FP16
```
[03/27/2025-10:01:30] [I] Throughput: 513.216 qps
[03/27/2025-10:01:30] [I] Latency: min = 1.85754 ms, max = 2.95117 ms, mean = 1.94697 ms, median = 1.94727 ms, percentile(90%) = 1.96875 ms, percentile(95%) = 1.98047 ms, percentile(99%) = 2.1084 ms
```

SiLU QAT (+38.18%)
```
[03/27/2025-09:18:52] [I] Throughput: 596.612 qps
[03/27/2025-09:18:52] [I] Latency: min = 1.63525 ms, max = 2.93359 ms, mean = 1.67465 ms, median = 1.67383 ms, percentile(90%) = 1.69141 ms, percentile(95%) = 1.69922 ms, percentile(99%) = 1.71094 ms
```

ReLU QAT (+90.62% from SiLU +60.36% from ReLU)
```
[03/27/2025-09:35:32] [I] Throughput: 822.998 qps
[03/27/2025-09:35:32] [I] Latency: min = 1.19727 ms, max = 2.46484 ms, mean = 1.21356 ms, median = 1.21094 ms, percentile(90%) = 1.21875 ms, percentile(95%) = 1.22266 ms, percentile(99%) = 1.23047 ms
```
